### DISCUSSION TITLE
can be given full solution "found" faster ?

### DISCUSSION THREAD

**Discussion by gregy4** [google/or-tools#3981] [CP-SAT questions]
Hello,

  I found noticeable difference in accepting of a solution between hints and equality constraints in a problem with around 200 thousand variables. The solution is complete as verified by equality constraints when all variables after presolve phase disappeared.

  With hints it takes about 400 seconds to confirm the solution, with equality constraints is everything completed in 4 seconds. Is it all probably about sat solver to process all variables or is it possible to "find" complete valid solution faster in a future ?

Thanks,
Jan

**Comment by sschnug** [comment]
Not an answer but some personal opinions:

- Fixing variables and hinting variables are completely different things and there are huge differences internally
  - In some sense this is related to the closed-world assumption and what is allowed to reason at which point in time
    - It's somewhat similar to assumption-based sat-solving (already introducing inefficiencies compared to variable-fixing)
    - Maybe one can also draw a connection to incremental-sat and the issue of some presolving-rules becoming unsound in some cases
      - We don't talk about incrementality here, but it's obvious that variable-fixing offers presolving which cannot be made while the domain of the variables is still a non-singleton -> yes... we seek a solution (all domains are singletons) and provide one (the singletone values); but we won't necessarily stop there and the solver must be ready to consider the remaining domain-values to further improve some objective (in some cases) which forbids presolving those away
- I think the solver will rewrite the model (let's make something up which sounds reasonably doable: decomposing some cumulative-constraint to multiple disjunctive/nooverlap-constraints)
  - I'm not so sure if there is always a 1:1 solution mapping possible without search (aka backtracking free)

But i'm also interested to hear some insights from the devs.

Edit: Further processing [Discussion 3945](https://github.com/google/or-tools/discussions/3945)
In some sense, every presolving-based reduction we lost due to not fixing variables (theoretical issues) will imply a need for propagation which will always cost time (without a very special fast-path for guaranteed singleton-domain hints). As those usually want to get "explained" i expect there lots of newly introduced variables (and mem allocs) to be needed. The question is: is there even something like a specialized fast-path possible or not.

**Reply by gregy4** [reply]
I agree with your arguments about direct usage of decisions made with fixed variables.
On other side I see fast way to verify validity of a solution with equality constraints. For valid solution a lot things can be done in advance - transfer objective bounds to search workers, run lns variants, transfer results between presolved cp models …

For this problem cp-sat solver can’t find feasible solution in 24hours and my alternative constraint solver finds first solution in a minute. There’s a question whether cp-sat approach is good enough for this size of problem or not when it takes more than 6x (c++ vs java) to find a solution with given recipe compared to a solver that really finds a solution of same problem.

**Reply by lperron** [reply]
please send us the model.

**Reply by gregy4** [reply]
Attached is a model with hints.

[model.zip](https://github.com/google/or-tools/files/13334241/model.zip)

**Reply by lperron** [reply]
your hint is not complete.

**Reply by lperron** [reply]
Meaning you hint 5556 variables out of 230k. So you need propagation of assign all the others, and return a feasible solution.

If I run your model on my workstation, with 9.8, the solver exits presolve after 75s, finds the first solution after 125s.

**Reply by gregy4** [reply]
Good point. The solution didn't contained all variables but rest of variables had no freedom in assignment since only one possible value was possible for each variable except variables in non-active optional intervals.

I tried to add explicitly hints for more variables, so far six iterations starting with around 73000 and ending with around 228000 hints so far. Comparison on or-tools 9.7 didn't showed noticeable progress with more hints.

Today I tried or-tools 9.8 with minimal and maximal number of hints (verified with fix_variables_to_their_hinted_value:true) on different os and computer with similar results.

Since a model contains a lot of optional intervals I tried to remove 50% of not active optional intervals (based on the solution hinted interval always preserved, 50% non-active intervals removed). Cp-sat solver found a solution with hints much faster. Of course models aren't directly comparable but it seems to me that the solver doesn't benefit from possibility to ignore non-active intervals for a solution based on hints.  

[model-hints6.zip](https://github.com/google/or-tools/files/13384555/model-hints6.zip)

**Comment by lperron** [comment]
One thing that can be done:
  - create the model with the partial hint
  - solve with the fix_variables_to_their_hinted_value:true
  - get the value of all variables in the solution
  - complete the hint with the complete solution and resolve with the parameter
